[CL-380] Rename Emails admin sidebar tab to Messaging

lilkraftwerk opened the pull request:
https://citizenlab.atlassian.net/browse/CL-380

Just renames the Emails tab to Messaging, without adding any new functionality. I'd like to make sure this is good and merge this one first before adding the `Texting` tab under a feature flag for the MVP.

Some of the messages have not been updated, like the subtitle on the messaging page `Send out your own emails and control what automated emails your users receive.` since that is still accurate as of now, when only emails are enabled. I left a comment below about one open question around translations.

Ideally we can have a couple people run this locally and make sure it all works properly. I checked on my machine and from what I can see it looks good but I'm relatively new and would like to be confident that a big change like this isn't breaking anything.

## Checklist

- [ ] Added entry to changelog
<details>
<summary>More info</summary>
Add a concise line to the 'Next release' section of the changelog (docs/README.md) so people other than developers can understand what has changed where. E.g. 'Added an error message to the project name field of the project edit form (Admin > Projects > Edit)'.
</details>

- [ ] WCAG 2.1 AA proof
<details>
<summary>More info</summary>
For front-end devs only. Is your work conforming with the WCAG 2.1 AA rules? If you need more info, read the [a11y page](https://www.notion.so/citizenlab/a11y-7568f83d42ab4895ac133b89d358997b) on our Notion.
</details>

- [ ] Tests
<details>
<summary>More info</summary>

### Unit tests

Did you add relevant unit tests?

### E2E tests

Sometimes it can be more efficient to update E2E tests after CI has run them. If you know which ones to update, go ahead! E2E template cl2-back:

```bash
docker compose run --rm web bin/rails cl2_back:create_tenant[localhost,e2etests_template]
```

</details>

- [ ] Prepared branch for code review
<details>
<summary>More info</summary>
Reviewed code to reduce unnecessary back and forth (removal of console.log, comments, ...)? Added comments to clarify code, emphasize what to pay attention to, etc.?
</details>

## Links

- [citizenlab-ee PR](**put URL here** or remove)
- [Specs](**put URL here** or remove)
- [Epic Deployment](**put URL here** or remove)

## How urgent is a code review?

Let the reviewer(s) know how urgent the code review is, so they can prioritize their work accordingly. Be specific (e.g. by Wednesday, end of the day/this week/... is better than 'urgent' or 'very urgent'). Optionally provide a word of explanation on your deadline.

lilkraftwerk committed:
[CL-380] Rename /emails routes and folders to /messaging/emails

Commit:
Translations updated by CI (extract-intl)

lilkraftwerk committed:
[CL-380] Update additional routes and copy for renamed Messaging tab

Commit:
Translations updated by CI (extract-intl)

lilkraftwerk committed:
[CL-380] Update additional messages from /emails to /messaging

lilkraftwerk committed:
[CL-380] Merge branch 'CL-380-fe-rename-emails-page-and-tabs-for-texting-mvp' of github.com:CitizenLabDotCo/citizenlab into CL-380-fe-rename-emails-page-and-tabs-for-texting-mvp

Commit:
Translations updated by CI (extract-intl)

lilkraftwerk committed:
[CL-380] Remove comment around translation

lilkraftwerk committed:
[CL-380] Merge branch 'CL-380-fe-rename-emails-page-and-tabs-for-texting-mvp' of github.com:CitizenLabDotCo/citizenlab into CL-380-fe-rename-emails-page-and-tabs-for-texting-mvp

lilkraftwerk commented:
As far as I can tell there's one remaining reference to the old URL, `admin/emails/custom`, in the translated key `engageArticle3Link`: 
![Screenshot 2022-03-14 at 09 04 37](https://user-images.githubusercontent.com/3614128/158130099-c1a5abf5-2d51-4e30-bf4d-1771f50d6c80.png)

I'm not sure that string actually needs to be translated, since it's not really copy but a URL, and in the current version there are no changes in any language:

![Screenshot 2022-03-14 at 09 04 43](https://user-images.githubusercontent.com/3614128/158130158-549c2123-fb73-444d-9bcb-2ec773f17586.png)

So we could potentially just hardcode the new URL in `EngageSection.tsx`, or if we want to continue using the translation system, I'm not 100% sure how to update that value. The [translations doc](https://www.notion.so/citizenlab/Translations-7e43f76d06a745b0b8455a0f7a27c9d1) says to change the translation key, but `engageArticle3Link` seems most accurate and matches the rest of the variable names in that file, so I wanted to check on what the best way to handle this is.

brentguf commented on front/cypress/integration/email_consent.ts:
```diff
@@ -1,7 +1,7 @@
   beforeEach(() => {
     cy.setAdminLoginCookie();
-    cy.visit('/admin/emails/custom/new');
+    cy.visit('/admin/messasing/emails/custom/new');
```
Typo? 

messasing => messaging

brentguf commented on front/app/containers/Admin/sideBar/index.tsx:
```diff
@@ -251,15 +251,15 @@ class Sidebar extends PureComponent<
-          link: '/admin/emails',
+          name: 'messaging',
+          link: '/admin/messaging',
           iconName: 'emails',
-          message: 'emails',
+          message: 'messaging',
           isActive: (pathName) =>
             pathName.startsWith(
               `${
                 getUrlLocale(pathName) ? `/${getUrlLocale(pathName)}` : ''
-              }/admin/emails`
+              }/admin/messaging`
             ),
         },
         {
```
Are we planning to use the email naming again later? Is messaging just for the index page?

lilkraftwerk replied:
Yeah, this is just the index, so `/admin/messaging` which will contain emails and texting once we're done. The previous routes were `/emails/custom` and `/emails/automated` and will now be `messaging/emails/custom` and `messaging/emails/automated`

jinjagit commented on front/cypress/integration/email_consent.ts:
```diff
@@ -1,7 +1,7 @@
   beforeEach(() => {
     cy.setAdminLoginCookie();
-    cy.visit('/admin/emails/custom/new');
+    cy.visit('/admin/messasing/emails/custom/new');
```
typo - spelling

lilkraftwerk replied:
Whoops. Good catch :)

jinjagit commented on front/app/translations/admin/ar-SA.json:
```diff
@@ -480,7 +480,6 @@
   "app.containers.Admin.emails.formUnexpectedErrorMessage": "حدث خطأ ما، يُرجى المحاولة مجددًا في وقت لاحق",
   "app.containers.Admin.emails.groups": "المجموعات",
   "app.containers.Admin.emails.helmetDescription": "أرسل الرسائل الإلكترونية إلى مجموعة مواطنين محددة وفعّل الحملات التلقائية",
-  "app.containers.Admin.emails.helmetTitle": "الرسائل الإلكترونية",
```
I get that there could be some translation changes, but I don't understand exactly what is being changed here and why (since I have never made such FE changes myself). I see deletions, and I guess these will need replacing. What is being deleted, and how will it be replaced?

lilkraftwerk replied:
This key is the `helmetTitle` for the old Emails tab. In practice it was used to set the window/tab name to `E-mails`, or its translated equivalent. Since that key was deleted and no longer exists in the codebase, our CI removes that key and value from all translation files, since it will never be accessed (check for commits from Circle-CI to see). 

The replacement is `"app.containers.Admin.messaging.helmetTitle": "Messaging"` to change the new tab title to `Messaging` and we'll need to wait for that to go through the Crowdin translation flow and be translated into languages other than English

jinjagit replied:
Thanks. So the new value 'Messaging' gets picked up from where?

brentguf replied:
A script that runs automatically: extract-intl.js

jinjagit replied:
@Brent. Yes, I assumed a process will catch the new value. But where does this process check for changed/missing values?

brentguf replied:
That's either front-build or front-extract-intl step in CI. They'll notify if something is missing (front-build) or when there are e.g. duplicates (front-extract-intl).

You can see them at the bottom of this PR.

jinjagit replied:
Yes, I am trying to work out what is the 'thing' it notices is missing. Looking at `extract-intl.js` I see filenames is one such thing. But I don't see any filename changes in this PR, so I am still looking to see how this works in this case.

brentguf replied:
I've personally never looked at these internals I think... The steps to change/update copy (without going into implementation) are in our Notion docs.

Why would you need to understand how that file works at this point?

lilkraftwerk committed:
[CL-380] Fix typo in messaging Cypress test

brentguf reviewed (commented):
+1 for smaller, focused PRs!

Regarding renaming, we use CamelCase for folders. So to be in line with the rest of the codebase, we need AutomatedEmails > emails-automated and CustomEmails > emails-custom.

jinjagit reviewed (approved):
LGTM 👍 

I tested manually, locally, and found no issues. Email campaigns seem to send OK, and the one automated email I tested also worked (assignment of Idea to an admin).

lilkraftwerk committed:
[CL-380] Rename folder to match other directory styles

lilkraftwerk committed:
[CL-380] Merge branch 'master' of github.com:CitizenLabDotCo/citizenlab into CL-380-fe-rename-emails-page-and-tabs-for-texting-mvp

brentguf committed:
Adjust old custom email URL

brentguf reviewed (changes requested):
There was one link that still needed to be changed. Can you -as a code review- check my latest commit?

engageArticle3Link in app/containers/Admin/guide/messages.ts needs to change as well. You need to change the key/id and then also adjust in the code.

This string seems not adjusted for any language (you can check this by looking at the JSON files with translations) so you don't need to add it on Crowdin and can just merge.

<img width="825" alt="Screenshot 2022-03-14 at 16 08 13" src="https://user-images.githubusercontent.com/16427929/158201451-a627348f-836a-494a-9f83-a78a703ffa67.png">

lilkraftwerk commented:
@brentguf your commit looks good. I left a comment about the final remaining translation key above (https://github.com/CitizenLabDotCo/citizenlab/pull/1623#issuecomment-1066489810) but it seems like the best way to do it would be creating a new key, to replace the existing one with the new `/messaging` prefaced URL?

lilkraftwerk committed:
[CL-380] Update engagement link to use new messaging namespace

lilkraftwerk committed:
[CL-380] Rename email folders to be more easily discoverable

brentguf commented:
> @brentguf your commit looks good. I left a comment about the final remaining translation key above ([#1623 (comment)](https://github.com/CitizenLabDotCo/citizenlab/pull/1623#issuecomment-1066489810)) but it seems like the best way to do it would be creating a new key, to replace the existing one with the new `/messaging` prefaced URL?

✅ yes.

I'll still inform you on how to deal with this translation-wise because I'm actually not sure about how we can do this without creating confusion.

Commit:
Translations updated by CI (extract-intl)

lilkraftwerk committed:
[CL-380] Update routes to match new folder structure

lilkraftwerk commented:
For the `engageArticle3Link` convo, I'm not sure what the best solution is. I hesitate to update all of these components to handle the single case we're talking about (the string is currently translated but doesn't need to be since the URL will never change) but the other option (sending this URL string through the translation flow unnecessarily) is also not great. Wrapping the messages in the `FormattedMessage` helper beforehand also seems more complex for anyone who reads this code later. 

Updating just the prop type definition to accept a plain string or `ReactIntl.FormattedMessage.MessageDescriptor` doesn't work since `FormattedMessage` accepts an object. Wrapping all of the props in the parent component with `FormattedMessage` seems tricky since we add HTML tags to them in the child component and this might be confusing in the future, or make component reuse trickier.

One solution I can think of is passing in an object like `{ id: "test", defaultMessage: "/messaging/admin/emails/custom" }` which is very much a hack but maybe the simplest way to do it, if we leave a comment as well. Not sure if that's the best solution but it requires the least amount of changes to the component setup here, which other than this one issue seems fine.

brentguf commented:
@lilkraftwerk adding string as a prop type, then doing a typeof === "string" check on this value in the component it's rendered. Would that work? If it's not a string, it's a message object, so then it should use formatMessage. This seems a decent solution, not introducing more debt but moving to the solution we rather want.

Wdyt?

lilkraftwerk committed:
[CL-380] Adjust AdminGuideArticle so a string URL can be passed

lilkraftwerk committed:
[CL-380] Add hardcoded custom email link to engage component

lilkraftwerk committed:
[CL-380] Clean up proptypes and INTL for AdminGuideArticle

lilkraftwerk committed:
[CL-380] Remove unused translation keys for engage links

lilkraftwerk committed:
[CL-380] Correct link for manual email engagement link

lilkraftwerk commented:
Cool, that makes sense. I changed all of the links to hardcoded strings (if you check the intl files, they're all the same in that they're all internal links with no translation in any locale) and updated the types to please the TS checker. Also removed the now-unused intl keys/values for `engageArticle_Link` 1-3.

Ok, all green

Commit:
Translations updated by CI (extract-intl)

lilkraftwerk committed:
[CL-380] Handle string or message object types in AdminGuideArticle

lilkraftwerk committed:
[CL-380] Merge branch 'CL-380-fe-rename-emails-page-and-tabs-for-texting-mvp' of github.com:CitizenLabDotCo/citizenlab into CL-380-fe-rename-emails-page-and-tabs-for-texting-mvp

brentguf commented on front/app/containers/Admin/routes.ts:
```diff
@@ -9,7 +9,7 @@ import invitationsRoutes from './invitations/routes';
 import settingsRoutes from './settings/routes';
 import settingsAreasRoutes from './settings/areas/routes';
-import emailsRoutes from './emails/routes';
+import messagingsRoutes from './messaging/routes';
```
Suggestion: would name messagingRoutes, but not important.

brentguf reviewed (approved):
👍

Lmk if you need help with the merge conflict.

lilkraftwerk committed:
[CL-380] Merge branch 'master' of github.com:CitizenLabDotCo/citizenlab into CL-380-fe-rename-emails-page-and-tabs-for-texting-mvp

lilkraftwerk committed:
[CL-380] Update changelog for renaming Emails tab

lilkraftwerk commented:
Just synced this with Crowdin, so waiting on translation branch to be complete before merging

lilkraftwerk committed:
[CL-380] Merge branch 'master' of github.com:CitizenLabDotCo/citizenlab into CL-380-fe-rename-emails-page-and-tabs-for-texting-mvp
